File: Algorithm Time Complexities/RecursiveBetterPowerXN.cpp
#include <bits/stdc++.h>
using namespace std;

int power(int x, int n){
    if (n == 0){
        return 1;
    }

    int smallPower = power(x, n/2);
    if (n % 2 == 0){
        return smallPower * smallPower;
    }else{
        return x * smallPower * smallPower;
    }
}

int main(){
    int x, n;
    cin >> x >> n;

    cout << power(x, n);
}

File: Algorithm Time Complexities/RecursiveBetterPowerXN.py
def power(x, n):
     
    if (n == 0):
        return 1
     
    smallPower = power(x, n//2)
    if (n % 2 == 0):
        return smallPower * smallPower
    else:
        return x * smallPower * smallPower
     
     
print(power(2, 3))

TC = O(logn)
SC = O(logn)
